File: plugin_sensie/ios/Classes/PluginSensiePlugin.m
#import "PluginSensiePlugin.h"
#import "FLRSynth.h"
#import <SensieFramework/SensieFramework-Swift.h>

@implementation PluginSensiePlugin{
  int _numKeysDown;
  FLRSynthRef _synth;
    Wrapper *instance = [[Wrapper alloc] init];
}

- (instancetype)init {
  self = [super init];
  if (self) {
    _synth = FLRSynthCreate();
    FLRSynthStart(_synth);
  }
  return self;
}

- (void)dealloc {
  FLRSynthDestroy(_synth);
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"plugin_sensie"
            binaryMessenger:[registrar messenger]];
  PluginSensiePlugin* instance = [[PluginSensiePlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call
                  result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS "
        stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else if ([@"onKeyDown" isEqualToString:call.method]) {
    FLRSynthKeyDown(_synth, [call.arguments[0] intValue]);
    _numKeysDown += 1;
    result(@(_numKeysDown));
  } else if ([@"onKeyUp" isEqualToString:call.method]) {
    FLRSynthKeyUp(_synth, [call.arguments[0] intValue]);
    
    _numKeysDown -= 1;
    result(@(_numKeysDown));
  } else if ([@"whipCounter" isEqualToString:call.method]) {
      NSNumber *result = [instance callGlobalWhipCounterWithParam: ];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end

File: plugin_sensie/ios/Classes/PluginSensiePlugin.swift
import Flutter
import UIKit
import SensieFramework

public class PluginSensiePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "com.sensie.plugin_sensie/swift_function", binaryMessenger: registrar.messenger())
    let instance = PluginSensiePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "whipCounter" {
      guard let args = call.arguments as? [String: Any] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        return
      }
      guard let unwrappedArg1 = args as? NSDictionary else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        return
      }
      let nativeResult = SensieFramework.whipCounter(param: unwrappedArg1)
      result(nativeResult)
    } else if call.method == "signalStrength" {
      guard let args = call.arguments as? [String: Any] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        return
      }
      let arg1 = args["sensies"] as? Array<Dictionary<String, Any>>
      guard let unwrappedArg1 = arg1 else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        return
      }
      let nativeResult = SensieFramework.signalStrength(sensies: unwrappedArg1)
      result(nativeResult)
    } else if call.method == "evaluateSensie" {
      guard let args = call.arguments as? [String: Any] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        return
      }
      let arg1 = args["sensie"] as? Dictionary<String, Any>
      let arg2 = args["sensies"] as? Array<Dictionary<String, Any>>
      guard let unwrappedArg1 = arg1 else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        return
      }
      guard let unwrappedArg2 = arg2 else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        return
      }
      let nativeResult = SensieFramework.evaluateSensie(sensie: unwrappedArg1, sensies: unwrappedArg2)
      result(nativeResult)
    } else {
      result(FlutterMethodNotImplemented)
    }
  }
}